src/include/simeng/OS/SimOS.hh

```diff
@@ -0,0 +1,137 @@
+#include "simeng/OS/SyscallHandler.hh"
+#include "simeng/SpecialFileDirGen.hh"
+#include "simeng/arch/Architecture.hh"
+#include "simeng/arch/aarch64/Architecture.hh"
+#include "simeng/arch/aarch64/Instruction.hh"
+#include "simeng/arch/riscv/Architecture.hh"
+#include "simeng/arch/riscv/Instruction.hh"
+#include "simeng/memory/SimpleMem.hh"
+#include "simeng/span.hh"
+
+// Forward declare RegressionTest class so that it can be declared a friend.
+class RegressionTest;
+
+namespace simeng {
+namespace OS {
+
+// Program used when no executable is provided; counts down from
+// 1024*1024, with an independent `orr` at the start of each branch.
+static uint32_t hex_[8] = {
+    0x320C03E0,  // orr w0, wzr, #1048576
+    0x320003E1,  // orr w0, wzr, #1
+    0x71000400,  // subs w0, w0, #1
+    0x54FFFFC1,  // b.ne -8
+                 // .exit:
+    0xD2800000,  // mov x0, #0
+    0xD2800BC8,  // mov x8, #94
+    0xD4000001,  // svc #0
+};
+
+/** A simple, lightweight Operating System kernel based on Linux to emulate
+ * syscalls and manage process execution. */
+class SimOS {
+ public:
+  /** Construct a SimOS object. */
+  SimOS(std::string executablePath, std::vector<std::string> executableArgs,
+        std::shared_ptr<simeng::memory::Mem> mem);
+
+  /** Tick SimOS. */
+  void tick();
+
+  /** Get a process with specified TID. */
+  const Process& getProcess(uint64_t TID) const;
+
+  /** Get shared_ptr to syscallHandler instance. */
+  std::shared_ptr<SyscallHandler> getSyscallHandler() const {
+    return syscallHandler_;
+  }
+
+  /** Register a core with the OS to enable process scheduling. */
+  void registerCore(std::shared_ptr<simeng::Core> core) {
+    cores_.push_back(core);
+  }
+
+  /** Check if OS has halted. */
+  bool hasHalted() const { return halted_; };
+
+  /** Set up friend class with RegressionTest to enable exclusive access to
+   * private functions. */
+  friend class ::RegressionTest;
+
+ private:
+  /** Create the initial SimOS Process running above this kernel from command
+   * line arguments.
+   * Empty command line arguments denote the usage of hardcoded instructions
+   * held in the hex_ array.*/
+  void createInitialProcess();
+
+  /** Construct the special file directory. */
+  void createSpecialFileDirectory() const;
+
+  /** The path of user defined Executable. */
+  std::string executablePath_ = DEFAULT_STR;
+
+  /** The runtime arguments of the user defined executable. */
+  std::vector<std::string> executableArgs_ = {};
+
+  /** The map of user-space processes running above the OS kernel.
+   * Key = process' TID
+   * Value = Shared pointer to process object with TID = key
+   */
+  std::unordered_map<uint64_t, std::shared_ptr<Process>> processes_ = {};
+
+  /** Queue of processes waiting to be scheduled. */
+  std::queue<std::shared_ptr<Process>> waitingProcs_ = {};
+
+  /** Queue of processes which are due to be scheduled */
+  std::queue<std::shared_ptr<Process>> scheduledProcs_ = {};
```

Comment:
Can this comment please be refactored to something more descriptive like:
"Queue of processes that have successfully sent an interrupt signal to a core and are waiting for the core to go into an idle state so as to be scheduled on it"

Reply:
Done